File: path.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jan 22 10:55:08 2019

@author: Victor
"""

class Path:
    def __init__(self,instance,path):
        self.instance = instance
        self.path = path
        
        self.nodes = []
        self.edges = []
        
        self.isValid = True
        self.dist = 0
        self.weight = 0        
        self.length = len(path)        
        
        for i in range(len(path)-1):
            
            if not instance.hasEdge(path[i],path[i+1]):
                self.isValid = False
                break
            
            dist = instance.edgeDist(path[i],path[i+1])
            self.edges += [[dist,(path[i],path[i+1]),0]]
            self.dist += dist
            
            weight = instance.nodeWeight(path[i])
            self.nodes += [[instance.ph[path[i]],weight,path[i],0]]
            self.weight += weight
            
        weight = instance.nodeWeight(path[-1])
        self.nodes += [[instance.ph[path[-1]],weight,path[i],0]]
        self.weight += weight
        
        # worst case scenario
        self.edges.sort(reverse=True)
        self.nodes.sort(reverse=True)
    
        d1 = instance.d1
        for i in range(len(self.edges)):
            edge = self.edges[i][1]
            if d1 > instance.D[edge[0]][edge[1]]:
                dev = instance.D[edge[0]][edge[1]]
                self.edges[i][2] = dev
                d1 -= dev
            else:
                dev = d1
                self.edges[i][2] = dev
                d1 -= dev
                break
    
        d2 = instance.d2
        for i in range(len(self.nodes)):
            if d2 > 2:
                dev = 2
                self.nodes[i][3] = dev
                d2 -= dev
            else:
                dev = d2
                self.nodes[i][3] = dev
                d2 -= dev
                break    
    
        self.worstDist = self.dist
        for dist, edge, dev in self.edges:
            self.worstDist += dist*dev
            
        self.worstWeight = self.weight
        for devWeight, weight, node, dev in self.nodes:
            self.worstWeight += devWeight*dev
        
    def __str__(self):
        s = "Path : " + str(self.path[0])
        for node in self.path[1:]:
            s += " -> "
            s += str(node)
        
        s += "\n"
        s += "Distance : " + str(self.dist) + " (" + str(self.worstDist) + ")\n"
        s += "Weight : " + str(self.weight) + " (" + str(self.worstWeight) + ")"
        
        return s

File: C++/path.cpp
#include <algorithm>
#include "path.h"

bool nodeCompare(PathNode node1, PathNode node2) {
	if (node1.devWeight > node2.devWeight) return true;
	else if (node1.devWeight < node2.devWeight) return false;
	else return node1.node > node2.node;
}

bool edgeCompare(PathEdge edge1, PathEdge edge2) {
	if (edge1.dist > edge2.dist) return true;
	else if (edge1.dist < edge2.dist) return false;
	else return edge1.node1 > edge2.node1;
}

Path::Path(const Instance& instance, vector<Node> path) :
instance(instance),
path(path)
{
	isValid = true;
	dist = 0;
	weight = 0;
	length = path.size();

	for (unsigned int i = 0; i < path.size()-1; ++i) {
		if (!instance.hasEdge(path[i],path[i+1])) {
			isValid = false;
			break;
		}

		int d = instance.edgeDist(path[i],path[i+1]);
		PathEdge edge = {path[i],path[i+1],d,0.};
		edges.push_back(edge);
		dist += d;

		int w = instance.nodeWeight(path[i]);
		PathNode node = {path[i],w,(float) instance.ph[path[i]],0.};
		nodes.push_back(node);
		weight += w;
	}

	int w = instance.nodeWeight(path.back());
	PathNode node = {path.back(),w,(float) instance.ph[path.back()],0.};
	nodes.push_back(node);
	weight += w;

	std::sort(nodes.begin(),nodes.end(),nodeCompare);
	std::sort(edges.begin(),edges.end(),edgeCompare);

	float d1 = (float) instance.d1;
	for (unsigned int i = 0; i < edges.size(); ++i) {
		float D = instance.D[edges[i].node1][edges[i].node2];
		if (d1 > D) {
			edges[i].dev = D;
			d1 -= D;
		}
		else {
			edges[i].dev = d1;
			d1 = 0;
			break;
		}
	}

	float d2 = (float) instance.d2;
	for (unsigned int i = 0; i < nodes.size(); ++i) {
		if (d2 > 2.) {
			nodes[i].dev = 2.;
			d2 -= 2.;
		}
		else {
			nodes[i].dev = d2;
			d2 = 0.;
			break;
		}
	}

	worstDist = dist;
	for (unsigned int i = 0; i < edges.size(); ++i) {
		worstDist += edges[i].dist * edges[i].dev;
	}

	worstWeight = weight;
	for (unsigned int i = 0; i < nodes.size(); ++i) {
		worstWeight += nodes[i].devWeight * nodes[i].dev;
	}

}
